<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="robots">
  <title>FEST analysis — User Manual</title>

</head>

<body style="background-color:#e6ffcc;">
<p><span style="font-size: 16pt; font-weight: 700"><a name="top"></a>How to use the FEST web application</span></p>
<p>This document contains the following sections:</p>

<ul>
  <li><a href="#Overview">Overview</a></li>
  <li><a href="#inputFiles">Input files</a></li>
  <li><a href="#inputParams">Input parameters</a></li>
  <li><a href="#outputParams">Output</a></li>
  <li><a href="#ref">Reference</a></li>
  <li><a href="#releases">News and releases</a></li>
</ul>

  <h2><a name="Overview">Overview</a></h2>
  <p>The <a href="http://www.stat-apps.onc.jhmi.edu/FEST/">FEST</a> (Functional Expansion of Specific T-cells) platform is a Shiny application for analysing T‑cell receptor (TCR) sequencing from short‑term, peptide‑stimulated cultures to identify antigen‑specific clonotypic expansions. The app loads TCR sequencing data (tab‑delimited), runs the analysis, and provides visualizations and downloadable results. By default the analysis is performed at the amino‑acid level (identical amino‑acid sequences are collapsed); you can select nucleotide‑level analysis with the <b>Use nucleotide level data</b> option on the <b>Run Analysis</b> tab.</p>

  <p>The app supports two experiment types: with and without replicates. For experiments without replicates FEST uses pairwise Fisher's exact tests to compare a culture of interest to a reference culture; for experiments with replicates it fits a negative‑binomial model across conditions. Each clone is tested separately; p‑values are adjusted using the Benjamini–Hochberg (FDR) procedure.</p>

  <p>The analysis is performed in the following steps:</p>
  <ol>
    <li><b>Load Data</b>: upload FEST files or a previously saved R object with data.</li>
    <li>After the data is successfully loaded, select the <b>Run Analysis</b> tab to run the analysis. Please make sure to specify whether the input data has replicates or not. <b>Important:</b> If the input data has replicates, the conditions will be extracted from the input file names. To be able to correctly extract the conditions, file names should follow the format <code>sampleID_condition_replicate.ext</code> (for example <code>sample1_HIV_1.tsv</code>).</li>
    <li>After the analysis is done, select the <b>Save Results</b> tab to save the results.</li>
  </ol>

  <p>To identify FEST‑expanded and FEST‑positive clonotypes you can apply thresholds on the <b>Save Results</b> tab. A clonotype is considered positive if:</p>
  <ol>
    <li>It is significantly expanded in the culture of interest versus the reference (FDR below the selected threshold). <b>Note:</b> this step is skipped if <b>Compare to reference</b> is disabled.</li>
    <li>It is significantly expanded in the culture of interest versus other cultures run in the experiment (this criterion is not used for FEST‑expanded clones).</li>
    <li>It has an odds ratio above the selected value.</li>
    <li>Its abundance (percent) exceeds the selected threshold.</li>
    <li>It has non‑zero abundance in at least the specified percent of conditions.</li>
  </ol>

  <p>The results from the same run can be saved several times with different thresholds without re‑running the analysis. Clones can also be filtered by minimal template counts (the <b>Specify the minimal number of templates</b> field on the <b>Run Analysis</b> tab). FEST‑expanded and FEST‑positive clones together with the run parameters are saved in the output Excel file (use the <b>Download Results</b> button on the <b>Save Results</b> tab). FEST‑positive clones can be visualized as a heat map (the <b>Create Heat Map</b> button).</p>

<p><a href="#top"><font size="2">Top of Page</font></a></p>

<h2><a name="inputFiles">Input files</a></h2>
<p>First, select the input file format. The app accepts two input types:</p>
<ul>
  <li>TCR sequencing data in tab‑delimited format (for example, exported from <a href="https://www.adaptivebiotech.com/immunoseq">Adaptive Biotechnologies ImmunoSEQ</a>, <a href="https://vdjtools-doc.readthedocs.io/en/master/intro.html">VDJtools</a>, or similar tools).</li>
  <li>An <code>.rda</code> file previously saved from this app.</li>
</ul>
<p><b>NOTE:</b> Multiple file selection is not supported on some older browsers (for example, Internet Explorer 9 and earlier).</p>
<p><a href="#top"><font size="2">Top of Page</font></a></p>

<h2><a name="inputParams">Input parameters</a></h2>
<p>After loading data, go to the <b>Run Analysis</b> tab and set the following parameters:</p>
<ul>
  <li><b>Analyze with replicates</b>. Check this when your experiment includes replicates. When enabled, the app extracts conditions from filenames — verify them in the main panel. Filenames should use the format <code>sampleID_condition_replicate.ext</code> (for example <code>sample1_HIV_1.tsv</code>).</li>
  <li><b>Use nucleotide level data</b>. Select this to perform analysis at the nucleotide level rather than the amino‑acid level. By default the analysis uses amino‑acid level data.</li>
  <li><b>Compare to reference</b>. Uncheck to omit the reference comparison step.</li>
  <li><b>Select a reference sample</b>. A reference sample is typically T cells cultured with cytokines but without peptide.</li>
  <li><b>Select sample(s) to exclude</b>. Excluded samples are omitted from analysis but information about them is included in the output file.</li>
  <li><b>Specify the minimal number of templates</b>. Minimum number of templates required to include a clone in the analysis.</li>
</ul>
<p>After setting parameters, press the <b>Run Analysis</b> button.</p>
<p><a href="#top"><font size="2">Top of Page</font></a></p>

<h2><a name="outputParams">Output</a></h2>
<p>After analysis completes, go to the <b>Save Results</b> tab and specify thresholds to determine FEST‑expanded and FEST‑positive clones. The odds ratio measures the magnitude of expansion of a clone in cultures of interest. The results are exported by clicking the <b>Download Results</b> button as an Excel workbook containing the following sheets:</p>
<ul>
  <li><b>condition_summary</b>: number of FEST‑positive clones and the sum of their percents per condition.</li>
  <li><b>positive_clones_summary</b>: list of FEST‑positive clones, the condition where each clone is expanded, per‑sample frequencies, and fold changes relative to baseline/reference where applicable.</li>
  <li><b>positive_clones_all_data</b>: extended information for FEST‑positive clones including template counts (<code>*_abundance</code> columns) and percents (<code>*_percent</code> columns) for all samples (including those excluded from the analysis).</li>
  <li><b>ref_comparison_only</b>: FEST‑expanded clones relative to the reference sample; includes FDR columns (<code>FDR:*</code>), odds ratios (<code>OR:*</code>), and abundance/percent columns for all samples. (This sheet is omitted if <b>Compare to reference</b> is disabled.)</li>
  <li><b>parameters</b>: all input parameters, number of samples analyzed, and total template counts for productive clones per sample.</li>
</ul>
<p>You can also create a heat map of FEST‑positive clones with the <b>Create Heat Map</b> button.</p>
<p><a href="#top"><font size="2">Top of Page</font></a></p>

<h2><a name="ref">Reference</a></h2>
<p>The app is publicly available at <a href="http://www.stat-apps.onc.jhmi.edu/FEST/">http://www.stat-apps.onc.jhmi.edu/FEST/</a>. The source code is on GitHub: <a href="https://github.com/OncologyQS/replicateFest">https://github.com/OncologyQS/replicateFest</a>.</p>

<p>The MANAFEST manuscript is published in <a href="http://cancerimmunolres.aacrjournals.org/content/early/2018/06/12/2326-6066.CIR-18-0129" target="_blank">Cancer Immunology Research</a>.</p>
<p>To cite the paper, please use the following reference:<br>
The Mutation-Associated Neoantigen Functional Expansion of Specific T cells (MANAFEST) assay: a sensitive platform for monitoring antitumor immunity<br>
Ludmila Danilova, Valsamo Anagnostou, Justina X Caushi, John-William Sidhom, Haidan Guo, Hok Yee Chan, Prerna Suri, Ada J. Tam, Jiajia Zhang, Margueritta El Asmar, Kristen A Marrone, Jarushka Naidoo, Julie R. Brahmer, Patrick M Forde, Alexander S. Baras, Leslie Cope, Victor E. Velculescu, Drew Pardoll, Franck Housseau and Kellie N. Smith<br>
Cancer Immunol Res June 12 2018 DOI: 10.1158/2326-6066.CIR-18-0129
</p>

<p>The version of the app published in Cancer Immunology Research has been deposited at <a href="https://sourceforge.net/projects/manafest/">https://sourceforge.net/projects/manafest/</a></p>

<p><a href="#top"><font size="2">Top of Page</font></a></p>

<h2><a name="releases">News and releases</a></h2>
<ul>
  <li><b>Version 15</b> (2025-06-10)
    <ol>
      <li>Separated load data, run analysis, and save results into different tabs.</li>
      <li>Added the "analysis with replicates" option.</li>
      <li>Increased the default for the number of reads to 50.</li>
      <li>Removed the baseline option.</li>
      <li>Switched to <code>openxlsx</code> for saving results to Excel.</li>
      <li>Save a list with analysis results and the input parameters used for the analysis (the <code>analysisRes</code> object) instead of only the analysis results.</li>
      <li>Switched to the <code>pheatmap</code> package for creating heatmaps.</li>
      <li>Added the threshold for the percent of non‑zero conditions.</li>
    </ol>
  </li>
  <li><b>Version 14</b> (2025-03-05) <a href=https://dobbs-onc-jhmi.shinyapps.io/Manafest_analysis_v14/> https://dobbs-onc-jhmi.shinyapps.io/Manafest_analysis_v14/</a>
    <ol>
      <li>Use the <code>immunarch</code> package for reading input tab‑delimited files.</li>
      <li>Split loading the data and analysis into different tabs.</li>
      <li>Added the percentage threshold.</li>
      <li>Fixed heatmap output when there is no or only one clone.</li>
      <li>Added all functions to the <code>replicateFest</code> package.</li>
    </ol>
  </li>
</ul>

<p><a href="#top"><font size="2">Top of Page</font></a></p>

</body>
</html>
