<!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
	<meta name="robots">
  <title>The FEST analysis user's manual</title>

</head>

<body style="background-color:#e6ffcc;">
<p><span style="font-size: 16pt; font-weight: 700"><a name="top"></a>How to work with the FEST web application</span></p>
<p>This article contains the following sections:</p>

<ul>
	<li><a href="#Overview">Overview</a></li>
	<li><a href="#inputFiles">Input files</a></li>
	<li><a href="#inputParams">Input parameters</a></li>
	<li><a href="#outputParams">Output</a></li>
	<li><a href="#ref">Reference</a></li>
	<li><a href="#releases">News and releases</a></li>	
</ul>

		<h2><a name="Overview">Overview</a></h2>
		<p>The <a href="http://www.stat-apps.onc.jhmi.edu/FEST/">FEST</a> (Functional Expansion of Specific T-cells) platform is 
			a Shiny application for analysis of TCR sequencing of short-term, 
		peptide-stimulated cultures to identify antigen-specific clonotypic amplifications. 
		The app loads TCR sequencing data in the tab-delimited format,
 performs the analysis, and visualizes and saves results. For analysis, we use only productive clones and
 summarize template counts for nucleotide sequences that translated into the same amino acid sequence.
 The analysis can be performed for two types of experiments - with and without replicates. The different tests are used for each type.
For experiments without replicates, we use the pairwise Fisherâ€™s exact test to compare the number of templates in a culture of interest		
 (with peptide) and a reference culture (T cells cultured with cytokines but without peptide). For experiments with replicates, we fit 
 the negative binomial model for all conditions. The analysis is done for each clone separately. The p-values are adjusted by Benjamini-Hochberg
 procedure (FDR). 
 <p>
	The analysis is performed in the follwoing steps:
   <ol>
     <li><b>Load data</b>: upload FEST files or a previously saved R
              object with data.</li>
      <li>After the data is successfully loaded,
              select the <b>Run analysis</b>
                   tab to run the analysis. Please make sure to specify
                   whether the input data has replicates or not.
                   <br><b>Important!</b> If the input data has replicates, the conditions
                   will be extracted from the input file names.
                   To be able to correctly extract the conditions, file names should
                   follow the format 'sampleID_condition_replicate.ext'.
                   E.g. 'sample1_HIV_1.tsv'.                   
      <li>After the analysis is done,
              select the <b>Save results</b> tab to save the results.</li>
	</ol>
    
 </p>
 
 To determine antigen-specific clonotypes (FEST positive clones), we use the following criteria specified on 
 the <b>Save results</b> tab, a clonotype: 
 <ol>
	 <li>should be significantly expanded in the culture of interest compared to the reference
	 culture at an FDR less than the specified threshold, <br>
	 <b>NOTE:</b> This step is omitted if the <b>Compare to reference</b> option is disabled. </li>
	 <li>should be significantly expanded in the culture of interest compared to every other culture
	 performed in tandem,</li> 
	 <li>should have an odds ratio above a specified value, </li>
	 <li>should have a percent above a specified value. </li>	 
 </ol>
 Clones can then be subject to a threshold for a minimal number of templates (the <b>Specify the minimal number of templates</b> field 
 on the <b>Run analysis</b> tab). FEST positive clones are saved in the output table (the <b>Download Results</b> button on 
 the <b>Save results</b> tab) and plotted as an output heat map (the <b>Create Heat Map</b> button).</p>

<a href="#top"><font size="2">Top of Page</font></a></P>

<h2><a name="inputFiles">Input files</a></h2>
<p> First, you need to specify a format of your input file(s). The app allows you to upload two type of data:
		<ul>
			<li>TCR sequencing data in the tab-delimited format exported from 
				<a href="https://www.adaptivebiotech.com/immunoseq">Adaptive Biotechnologies ImmunoSEQ</a> platform, 
				 <a href="https://vdjtools-doc.readthedocs.io/en/master/intro.html">VDJtools</a>, and others.
			</li>
			<li>or .rda file that were previously saved from the app</li>
			<b>NOTE:</b> Multiple file selection does not work on older browsers, including Internet Explorer 9 and earlier.
		</ul>
		</p>
		<a href="#top"><font size="2">Top of Page</font></a>

<h2><a name="inputParams">Input parameters</a></h2>
	<p>After loading the data, go to the <b>Run Analysis</b> tab and specify the following input parameters:
	
	<ul>
		<li> <b>The input with replicates</b>. If your experiment has replicates.</li>
		<li> <b>Compare to reference</b>. If you want to omit the comparison to a reference sample step, disable this option.</li>
		<li> <b>Use nucleotide level data</b>. Select this option if you want to perform analysis on the nucleotide level data instead of amino acid level data. 
			By default, the analysis is performed on amino acid level.</li>
		<li> <b>Select a reference sample</b>. It is usually T cells cultured with cytokines but without peptide.</li>
		<li> <b>Select sample(s) to exclude</b>. The samples will be excluded from analysis, but information about 
			them will be included in the output file</li>
		<li> <b>Specify the minimal number of templates</b>. A threshold for the number of templates in cultures of interest</li>
	</ul>
	After setting all parameters, press the <b>Run Analysis</b> button.
	
	</p>
	<a href="#top"><font size="2">Top of Page</font></a>

<h2><a name="outputParams">Output</a></h2>
		<p>	After the analysis is done, go to the <b>Save results</b> tab and specify FDR, odds ratio, and percentage thresholds 
			that will be used to determine FEST positive clones. 
		The positive odds ratio corresponds to the expansion of a clone in cultures of interest. 
		The results are saved by clicking the <b>Download Results</b> button in an Excel file with the following spreadsheets:
		<ul>
			<li> The <b>condition_summary</b> sheet contains the number of positive clones and the sum of percents of the clones per condition </li>
			<li> The <b>positive_clones_summary</b> sheet contains a list of positive clones, condition in which a clone is expanded, percent of a clone 
				in the baseline and the reference samples, and the fold change of a clone comparing to the baseline and the reference samples</li>
			<li> The <b>positive_clones_all_data</b> sheet contains extended information about the positive clones including 
			the number of templates (the '*_abundance' columns) and the percent (the '*_percent' columns)
			in all samples including those that were excluded from analysis 

			<li> The <b>ref_comparison_only</b> sheet contains a list of all analyzed clones, FDRs (the 'FDR:*' columns) and odds ratios (the 'OR:*' columns) for comparisons between cultures of interest 
				and the reference sample, as well as the number of templates (the '*_abundance' columns) and the percent (the '*_percent' columns)
				in all samples including those that were excluded from analysis<br>
				<b>NOTE:</b> This sheet is omitted if the <b>Compare to reference</b> option is disabled</li>
			<li> The <b>parameters</b> sheet contains all input parameters, the number of samples analyzed, and the total number of templates 
			corresponding to productive clones per sample</li>
		</ul>
		You also can save the positive clones as a heat map by clicking the <b>Create Heat Map</b> button.
		</p>
	<a href="#top"><font size="2">Top of Page</font></a></P>

<h2><a name="ref">Reference</a></h2>
	<p>The manuscript is published in <a href="http://cancerimmunolres.aacrjournals.org/content/early/2018/06/12/2326-6066.CIR-18-0129" target="_blank" >Cancer Immunology Research.</a></p>
	To cite the paper, please use the following reference:<br>
	The Mutation-Associated Neoantigen Functional Expansion of Specific T cells (MANAFEST) assay: a sensitive platform for monitoring antitumor immunity<br>
Ludmila Danilova, Valsamo Anagnostou, Justina X Caushi, John-William Sidhom, Haidan Guo, Hok Yee Chan, Prerna Suri, Ada J. Tam, Jiajia Zhang, Margueritta El Asmar, Kristen A Marrone, Jarushka Naidoo, Julie R. Brahmer, Patrick M Forde, Alexander S. Baras, Leslie Cope, Victor E. Velculescu, Drew Pardoll, Franck Housseau and Kellie N. Smith
<br>Cancer Immunol Res June 12 2018 DOI: 10.1158/2326-6066.CIR-18-0129
<p></p>
	<p>This shiny app is publicly available at <a href="http://www.stat-apps.onc.jhmi.edu/FEST/">http://www.stat-apps.onc.jhmi.edu/FEST/</a>
and the current version of is available at <a href="https://github.com/OncologyQS/replicateFest>	and the current version is available at GitHub <a href="https://github.com/OncologyQS/replicateFest">
		https://github.com/OncologyQS/replicateFest</a>.</p>

		<p>The older version of the app has been deposited at <a href="https://sourceforge.net/projects/manafest/">https://sourceforge.net/projects/manafest/</a></p>

	<a href="#top"><font size="2">Top of Page</font></a></P>

	<h2><a name="releases">News and releases</a></h2>
	<p>	<ul>
		<li> <b>Version 15</b> (2025-06-10)
			<ol>
				<li>separated load data, run analysis, and save the results to the different tabs</li>
				<li>added analysis with replicates</li>
				<li>increased the default for the number of reads to 50</li>
				<li>removed the baseline option</li>
				<li>switched to openxlsx for saving results into an Excel file</li>
				<li>made a list with analysis results and parameters that were used for the analysis 
					insted of just analysis results (the analysisRes object)</li>				
			</ol>	
		</li><li> <b>Version 14</b> (2025-03-05) 
			<a href=https://dobbs-onc-jhmi.shinyapps.io/Manafest_analysis_v14/> https://dobbs-onc-jhmi.shinyapps.io/Manafest_analysis_v14/</a>
			<ol>
				<li>use the immunarch package for reading input tab-delimited files</li>
				<li>split loading the data and analysis to different tabs</li>
				<li>added percentage threshold to the version without replicates</li>
				<li>fixed output to heatmap when there is no or one clone</li>
				<li>added all functions to the replicateFest package</li>
			</ol>	
		</li>
	</ul>
	<a href="#top"><font size="2">Top of Page</font></a></P>

	
</body>
</html>
